### ISSUE TITLE
Azure Storage integration

### ISSUE THREAD

**Issue by v-a-s-a** [ropensci/targets#864]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] New features take time and effort to create, and they take even more effort to maintain. So if the purpose of the feature is to resolve a struggle you are encountering personally, please consider first posting a "trouble" or "other" issue so we can discuss your use case and search for existing solutions first.
* [x] Format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Proposal

I've seen hints (e.g. in #769) at this while reading through the saga of GCP integration. For us, the day has come — we have a workflow running on Azure that we are trying to port to `targets`.

I'd be happy to start just as @MarkEdmondson1234 did in #722, and produce Azure Storage utility functions based on the existing AWS S3 functions. I can also look at the tests, and try to replicate a similar harness for Azure. Based on discussion in #720, it looks like there is then work to be done integrating these into the `targets` class hierarchy.

How does this sound for an initial plan to PR?

**Comment by wlandau** [comment]
Yes, I am happy to support Azure storage if you would be willing to help with the utilities. I am on vacation until July 20, and I will be able to work through the technical details and PRs with you then. Looking forward to it.

**Comment by v-a-s-a** [comment]
Thanks Will — enjoy the time away!

**Comment by wlandau** [comment]
@noamross mentioned http://min.io/ (example: https://github.com/noamross/targets-minio-versioning), which claims to support Azure. @v-a-s-a, is this a reasonable workaround for you? It would be nice to try this before implementing a native solution in `targets`.

**Comment by v-a-s-a** [comment]
I've submitted an intiall pull request with a test and some utilities. It took a bit of time to decide how to to authentication, and I settled on using a service principal (similar to GCP's service account). I've been testing using Azure's free $200 tier on a private account.

It's a bare start, but I thought I would get something into the pull request to start. It does look very similar to GCP, and it would have been great if we could use a single API to handle both.

Take a look at the initial setup and test and let me know what you think. If it's too much to integrate into targets directly, perhaps we can do a separate package, similar to `gittargets`?

Otherwise, I'll continue trying to replicate the work done with GCP. Following the changes there has been helpful. The `AzureR` family of packages is decent, and almost all the functionality —e.g. file versioning and meta-data — we need is exposed there.

**Comment by JamesHWade** [comment]
I'm happy to help, too. Azure + targets would make my work life a lot easier.

**Comment by wlandau** [comment]
Thank you both. As @v-a-s-a said at the top, it would be great to have Azure storage utilities just like the ones for [AWS](https://github.com/ropensci/targets/blob/main/R/utils_aws.R) and [GCP](https://github.com/ropensci/targets/blob/main/R/utils_gcp.R) (using `AzureAuth` and `AzureStor` ?). A PR would really help.

I am naive to Azure Storage, and I am wondering about a few specifics:

* For blob storage (I assume that's what we'll use, unless file storage is better), can we rely on [strong read-after-write consistency](https://aws.amazon.com/blogs/aws/amazon-s3-update-strong-read-after-write-consistency/)? 
* Would anything special need to happen for large files (> 5GB) or multi-part uploads? @MarkEdmondson1234's packages handle this automatically, but for AWS the [solution was manual](https://github.com/ropensci/targets/blob/90d81406c644954820cd37c11a8134161e5fa98f/R/utils_aws.R#L116-L203).

**Comment by alexkyllo** [comment]
> I am naive to Azure Storage, and I am wondering about a few specifics:
> 
>     * For blob storage (I assume that's what we'll use, unless file storage is better), can we rely on [strong read-after-write consistency](https://aws.amazon.com/blogs/aws/amazon-s3-update-strong-read-after-write-consistency/)?

According to [this](https://azure.microsoft.com/en-us/blog/managing-concurrency-in-microsoft-azure-storage-2/), Azure Blob Storage "was designed to embrace a strong consistency model which guarantees that when the Storage service commits a data insert or update operation all further accesses to that data will see the latest update." So that's good news.
 
>     * Would anything special need to happen for large files (> 5GB) or multi-part uploads? @MarkEdmondson1234's packages handle this automatically, but for AWS the [solution was manual](https://github.com/ropensci/targets/blob/90d81406c644954820cd37c11a8134161e5fa98f/R/utils_aws.R#L116-L203).

For Block Blobs the {AzureStor} package uses the PUT blocklist operation: [blob_upload_internal.R#L1](https://github.com/Azure/AzureStor/blob/962edb6ae89f598e3c5771236187c48bf808d82a/R/blob_upload_internal.R#L1) and according to [this documentation](https://learn.microsoft.com/en-us/azure/storage/blobs/scalability-targets#scale-targets-for-blob-storage) the max size for an upload using PUT blocklist is about 190.7 TiB, so I don't believe we'll need to implement any special handling for large files.

**Comment by wlandau** [comment]
This issue has been inactive for a few months, and I was planning to crowd-source key parts of the implementation, so I am converting it to a discussion.
